### PR TITLE
feat(cache-annotation): Move from Doctrine cache to Symfony cache

### PR DESCRIPTION
**Pull request by Shivoham**
https://simple-it.atlassian.net/browse/OC-46675

### CODE REVIEW

File: phpstan.neon [hunk 1]
```diff
@@ -29,3 +28,4 @@ parameters:
             - JsonException
             - PHPUnit\Framework\Exception
             - ReflectionException
+            - Psr\Cache\InvalidArgumentException
```
Comment (sidux) [inline phpstan.neon:31]: Shouldn't we catch this in the sf handler and rethrow a generic exception ? (or just forward it and declare it in the contract)

File: src/FrameworkBridge/Symfony/Config/services.php [hunk 2]
```diff
@@ -22,6 +23,11 @@
         \dirname(__DIR__, 4) . '/src/Handler/Handler/*'
     );
 
+    $services->set(SymfonyCacheHandler::class)
```
Comment (sidux) [inline src/FrameworkBridge/Symfony/Config/services.php:26]: why is this needed ? it will be detected and declared automatically I guess

File: phpstan.neon (hunk 1 above)
Comment (Shivoham) [reply phpstan.neon:31]: PSR's InvalidArgumentException is an interface that does not extend throwable (at least [in its 1.0 version](https://github.com/php-fig/cache/blob/1.0.1/src/InvalidArgumentException.php)), thus cannot be properly catched nor declared as a `@throws` in phpdoc without making phpstan yell.
I'm not sure how to handle this case without dropping support for symfony 3.4's components

File: src/FrameworkBridge/Symfony/Config/services.php (hunk 2 above)
Comment (Shivoham) [reply src/FrameworkBridge/Symfony/Config/services.php:26]: Symfony users may encounter autowiring error if they don't have an instance of TagAwareAdapterInterface declared in their services. 
I was assuming it is their concern to declare the handler they want to use, and then let the compiler pass do the rest of the job

File: src/Interceptor/Interceptor/CacheInterceptor.php [hunk 3]
```diff
@@ -98,44 +91,47 @@ public function getSuffixPriority(): int
...
-    private function getNamespace(Instance $instance, Cache $annotation): ?string
+    private function getNamespace(Instance $instance, Cache $annotation): string
     {
         $parameters = $instance->getMethod()
             ->getParameters()
         ;
+
         if ($annotation->getNamespace() !== null) {
-            $resolvedExpression = $this->resolveExpression(
+            return $this->resolveExpression(
```
Comment (sidux) [inline src/Interceptor/Interceptor/CacheInterceptor.php:101]: we are missing the `md5` here

File: src/FrameworkBridge/Symfony/Config/services.php (hunk 2 above)
Comment (sidux) [reply src/FrameworkBridge/Symfony/Config/services.php:26]: I see let's keep it like this for now, maybe in the future we'll handle a more advanced auto discovery

File: src/Interceptor/Interceptor/CacheInterceptor.php (hunk 3 above)
Comment (sidux) [reply src/Interceptor/Interceptor/CacheInterceptor.php:101]: or we can accept it (reloading all website cache at once), but removing md5 from everywhere

File: phpstan.neon (hunk 1 above)
Comment (sidux) [reply phpstan.neon:31]: we can drop support for symfony 3 and 4 since they are not supported anymore

Comment (Shivoham) [reply phpstan.neon:31]: Ok, after further investigations, psr/cache ^2.0 [requires at least php 8](https://github.com/php-fig/cache/blob/2.0.0/composer.json#L13)

Comment (sidux) [reply phpstan.neon]: Ok, we should remember to remove this after php 8 update

File: src/Interceptor/Interceptor/CacheInterceptor.php [hunk 4]
```diff
@@ -83,9 +77,18 @@ public function supportsSuffix(Instance $instance): bool
...
-            ->hasAnnotation(Cache::class)
-        ;
+        if (!$instance->getMethod()->hasAnnotation(Cache::class)) {
+            return false;
+        }
+
+        $method = $instance->getMethod();
+        $annotation = $method->getAnnotation(Cache::class);
+
+        if ($annotation->getHandler() === 'legacy_redis') {
```
Comment (sidux) [inline src/Interceptor/Interceptor/CacheInterceptor.php:87]: there is legacy redis and legacy array, we should only check that it starts with legacy

Comment (Shivoham) [reply src/Interceptor/Interceptor/CacheInterceptor.php:87]: indeed, good catch

File: tests/Double/Stub/Cache/LegacyCacheAnnotatedClass.php [hunk 5]
```diff
@@ -77,23 +77,23 @@ public function cacheWithIdAndParameters(ParameterClassStub $param1, $param2): s
...
+     * @Cache(handler="legacy_array", namespace="'test-namespace'", id="'toto'")
      */
     public function cacheWithNamespaceAndId(): string
     {
         return self::DATA;
     }
 
     /**
-     * @Cache(handler="legacy_redis", namespace="'test-namespace' ~ param1.publicField")
+     * @Cache(handler="legacy_array", namespace="'test-namespace' ~ param1.publicField")
```
Comment (sidux) [inline tests/Double/Stub/Cache/LegacyCacheAnnotatedClass.php:96]: we should maybe load the same handler with a different name to have a test for something other than `legacy_array`

**Review by sidux — APPROVED** [review]
